src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -565,16 +564,18 @@ bool MetaspaceShared::link_class_for_cds(InstanceKlass* ik, TRAPS) {
 
 void MetaspaceShared::link_and_cleanup_shared_classes(TRAPS) {
   // Collect all loaded ClassLoaderData.
-  CollectCLDClosure collect_cld;
+  ResourceMark rm;
+  GrowableArray<ClassLoaderData*> loaded_cld;
+  CollectCLDClosure collect_cld(&loaded_cld);
```

Comment:
I think we should add a comment to say why it's necessary to first collect the ClassLoaderDatas first:

```
// ClassLoaderDataGraph::loaded_cld_do requires ClassLoaderDataGraph_lock.
// We cannot link the classes while holding this lock (or else we may run into deadlock).
// Therefore, we need to first collect all the CLDs, and then link their classes after
// releasing the lock.
```